### ISSUE TITLE
Inconsistency of output type of `marginals` 

### ISSUE THREAD

**Issue by theogf** [cscherrer/MeasureBase.jl#33]
Not sure if this is a bug I find the behavior a bit inconsistent:
```julia
d1 = For(10) do _; Normal(); end;
d2 = For(rand(10)) do s; Normal(0,s); end;
d3 = For(rand(10), rand(10)) do m,s; Normal(m, s); end;
marginals(d1) isa AbstractVector # true (this is a MappedArray)
marginals(d2) isa AbstractVector # true (this is a MappedArray)
marginals(d3) isa AbstractVector # false (this is a Generator)
# And yet
size(d1) == size(d2) == size(d3)
```

**Comment by theogf** [comment]
I see now that `marginals` returns a `mappedarray` only if the arguments for `f` are `Arrays` themselves.
And the easy solution is just to create a closure on `m` and `s` as :
```
m = rand(10)
s = rand(10)
d3 = For(1:10) do i; Normal(m[i], s[i]); end;
```

**Comment by cscherrer** [comment]
Good catch. There's a branch called `dev2` where this is fixed:
```julia
julia> d1 = For(3) do _; Normal(); end
For{Normal{(), Tuple{}}}(#unused#->Main.Normal(), Base.OneTo(3))

julia> d2 = For(rand(3)) do s; Normal(0,s); end
For{Normal{(:μ, :σ), Tuple{Int64, Float64}}}(s->Main.Normal(0, s), [0.7440281294464448, 0.5332840336937033, 0.3985027428918755])

julia> d3 = For(rand(3), rand(3)) do m,s; Normal(m, s); end
For{Normal{(:μ, :σ), Tuple{Float64, Float64}}}((:m, :s)->Main.Normal(m, s), [0.5753478785718185,
                                                                             0.17230385216534816,
                                                                             0.08900389248766771], [0.9123088121937831,
                                                                                                    0.8654744479522047,
                                                                                                    0.6820302137885231])

julia> marginals(d1)
3-element mappedarray(#unused#->Main.Normal(), ::OneTo{Int64}) with eltype Normal{(), Tuple{}}:
 Normal()
 Normal()
 Normal()

julia> marginals(d2)
3-element mappedarray(s->Main.Normal(0, s), ::Vector{Float64}) with eltype Normal{(:μ, :σ), Tuple{Int64, Float64}}:
 Normal(μ = 0, σ = 0.744028)
 Normal(μ = 0, σ = 0.533284)
 Normal(μ = 0, σ = 0.398503)

julia> marginals(d3)
3-element mappedarray((:m, :s)->Main.Normal(m, s), ::Vector{Float64}, ::Vector{Float64}) with eltype Normal{(:μ, :σ), Tuple{Float64, Float64}}:
 Normal(μ = 0.575348, σ = 0.912309)
 Normal(μ = 0.172304, σ = 0.865474)
 Normal(μ = 0.0890039, σ = 0.68203)
```

BTW the story there is that I had thought `dev` would lead to the next release, but a conversation with @mschauer and @oschulz helped me see that some things were out of place. `dev2` started as an exploratory branch, but now things are really coming together. I think I'll merge that into `dev` soon so the naming isn't so weird.

BTW `For` usually holds things that look like a MappedArray. You could also have this as a `ProductMeasure{<:MappedArray}`, but I wanted 
1. A better way for users to specify this (short names for common things, and no need to wrap)
2. A nicer `show` representation
3. Potential to later expand with e.g. InfiniteArrays.jl or ContinuumArrays.jl

**Comment by theogf** [comment]
I think this is closed with 0.6 :)
